src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -7159,5 +7431,842 @@ void ExceptionTracker::ResetThreadAbortStatus(PTR_Thread pThread, CrawlFrame *pC
 }
 #endif //!DACCESS_COMPILE
 
-#endif // FEATURE_EH_FUNCLETS
+#ifndef DACCESS_COMPILE
+// Mark the pinvoke frame as invoking RhpCallCatchFunclet (and similar) for collided unwind detection
+void MarkInlinedCallFrameAsFuncletCall(Frame* pFrame)
+{
+    _ASSERTE(pFrame->GetVTablePtr() == InlinedCallFrame::GetMethodFrameVPtr());
+    InlinedCallFrame* pInlinedCallFrame = (InlinedCallFrame*)pFrame;
+    pInlinedCallFrame->m_Datum = (PTR_NDirectMethodDesc)((TADDR)pInlinedCallFrame->m_Datum | (TADDR)InlinedCallFrameMarker::ExceptionHandlingHelper | (TADDR)InlinedCallFrameMarker::SecondPassFuncletCaller); 
+}
+
+// Mark the pinvoke frame as invoking any exception handling helper
+void MarkInlinedCallFrameAsEHHelperCall(Frame* pFrame)
+{
+    _ASSERTE(pFrame->GetVTablePtr() == InlinedCallFrame::GetMethodFrameVPtr());
+    InlinedCallFrame* pInlinedCallFrame = (InlinedCallFrame*)pFrame;
+    pInlinedCallFrame->m_Datum = (PTR_NDirectMethodDesc)((TADDR)pInlinedCallFrame->m_Datum | (TADDR)InlinedCallFrameMarker::ExceptionHandlingHelper);
+}
+
+extern "C" void QCALLTYPE RhpAppendExceptionStackFrame(QCall::ObjectHandleOnStack exceptionObj, SIZE_T ip, SIZE_T sp, int flags, ExInfo *pExInfo)
+{
+    QCALL_CONTRACT;
+    
+    BEGIN_QCALL;
+    GCX_COOP();
+
+    Thread* pThread = GET_THREAD();
+    Frame* pFrame = pThread->GetFrame();
+    MarkInlinedCallFrameAsFuncletCall(pFrame);
+
+    bool canAllocateMemory = !(exceptionObj.Get() == CLRException::GetPreallocatedOutOfMemoryException()) &&
+                             !(exceptionObj.Get() == CLRException::GetPreallocatedStackOverflowException());
+
+    EECodeInfo codeInfo(ip);
```

Comment:
The stackwalker must have resolved IP to MethodDesc already. Is this redundant lookup leaving perf on the table?

Reply:
I can extract the `MethodDesc` from the passed in `ExInfo`.

Reply:
Done